keps/sig-apps/3715-indexed-job-with-unset-completions/README.md

```diff
@@ -0,0 +1,896 @@
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Currently Jobs are validated to have `spec.completions` set when `spec.completionMode`=`Indexed`. This KEP proposes to relax this validation and allow 
+Indexed jobs to be created with zero `spec.completions`. 
+
+When `spec.completions` is not set for an `Indexed` Job, the behavior is summarized as follows:
+- The success, failure and finish semantics are same as `NonIndexed`. Specifically:
+  - Success semantics:
+    - The job succeeded if any pod exits successfully. 
+    - The remaining active pods are allowed to continue to run to completion, and the job is marked as finished only 
+      when all pods complete. The success status doesn't change if any of the remaining pods fail, and they will not be retried.
+  - Failure semantics: 
+    - The job failed if the number of failures according to failure policy exceeds `spec.backofflimit`.
+- The range of indexes will be from 0 to `spec.parallelism`.
+- When `spec.parallelism` is updated to a lower value, the pods with higher indexes will be removed first (i.e, similar to StatefulSets)
+
+
+
+## Motivation
+
+There are cases where a batch workload requires an autoscaled indexed job with stable DNS pod names, for example MPI Horovord, Ray, PyTorch etc. 
+By modeling these workloads as an `Indexed` Job without `spec.completions` set (which is an immutable field), `spec.parallelism`, a mutable field, can be used to scale the Job. 
+While such workloads could be modeled as a StatefulSet, the Job API offers batch-specific features that makes it a better fit such as [pod and container failure handling](https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures)
+
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Allow `Indexed` Jobs to be created when `spec.completions` is not set.
+- Add special treatment for such jobs when `spec.parallelism` is scaled down by removing higher indexes first.
```

Comment:
Actually, this is not special behavior. Once we deduct indexes from parallelism, this will start happening.

So the goal is just to take the number of indexes from parallelism.

Reply:
On scale down, the list of pods to remove are ordered based on a heuristic that doesn't take the index into account, so it is a new behavior.

Here is where the job controller selects the candidates: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L1515

and here where the order func: https://github.com/kubernetes/kubernetes/blob/7b01daba714514aef68d501dc181cec946732481/pkg/controller/controller_utils.go#L716

Reply:
ok, that function name is `appendDuplicatedIndexPodsForRemoval`, so I thought it only removes "duplicate" indexes. I will update the sentence.